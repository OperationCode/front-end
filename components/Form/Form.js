import { connect } from 'formik';

export const Form = connect(({ formik: { handleReset, handleSubmit }, ...props }) => (
  <form onReset={handleReset} onSubmit={handleSubmit} noValidate {...props} />
));

Form.displayName = 'Form';
