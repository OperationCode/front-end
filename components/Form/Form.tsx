import { connect } from 'formik';
import type { DetailedHTMLProps, FormHTMLAttributes, FunctionComponent } from 'react';

const FormikConnectedForm = connect(({ formik: { handleReset, handleSubmit }, ...props }) => (
  <form onReset={handleReset} onSubmit={handleSubmit} noValidate {...props} />
)) as FunctionComponent<DetailedHTMLProps<FormHTMLAttributes<HTMLFormElement>, HTMLFormElement>>;

FormikConnectedForm.displayName = 'FormikConnectedForm';

export default FormikConnectedForm;
