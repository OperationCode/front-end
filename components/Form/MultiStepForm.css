.MultiStepForm {
  width: auto;
}

.containerFormErrorMessage {
  display: flex;
  justify-content: center;
}

.buttonGrouping {
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-around;
  align-items: center;
  margin-left: auto;
  margin-right: auto;
  margin-top: 2rem;

  /*
   * Specific value just so outer edges of each button visually aligns with
   * the outer edges of inputs in our often-existing two-column forms
   */
  width: 92.5%;
}

.buttonGrouping button + button {
  margin-left: 1rem;
}

@media screen and (--small-viewport) {
  .buttonGrouping button {
    min-width: 100px;
  }
}

.forceButtonRight {
  justify-content: flex-end;
}
