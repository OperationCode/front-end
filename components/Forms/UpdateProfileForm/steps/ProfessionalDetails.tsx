import { Field } from 'formik';
import * as Yup from 'yup';
import { validationErrorMessages } from 'common/constants/messages';
import Input from 'components/Form/Input/Input';
import { SelectSingle } from 'components/Form/Select/SelectSingle';
import { mapStringsToSelectOptions } from 'common/utils/array-utils';
import styles from './_steps.module.css';

ProfessionalDetails.title = 'Professional Details';

ProfessionalDetails.validationSchema = Yup.object().shape({
  employmentStatus: Yup.string().nullable().required(validationErrorMessages.required),
  companyName: Yup.string().nullable(),
  companyRole: Yup.string().nullable(),
});

ProfessionalDetails.initialValues = {
  employmentStatus: '',
  companyName: '',
  companyRole: '',
};

export type ProfessionalDetailsFormShape = typeof ProfessionalDetails.initialValues;

interface ProfessionalDetailsProps {
  isSubmitting: boolean;
}

const options = mapStringsToSelectOptions(['Full-Time', 'Part-Time', 'Unemployed']);

export function ProfessionalDetails({ isSubmitting }: ProfessionalDetailsProps) {
  return (
    <div className={styles.row}>
      <Field
        className={styles.fullWidth}
        name="employmentStatus"
        label="Employment Status*"
        component={SelectSingle}
        options={options}
        isDisabled={isSubmitting}
      />

      <Field
        type="text"
        name="companyName"
        label="Company Name"
        component={Input}
        isDisabled={isSubmitting}
      />

      <Field
        type="text"
        name="companyRole"
        label="Company Role"
        component={Input}
        isDisabled={isSubmitting}
      />
    </div>
  );
}
