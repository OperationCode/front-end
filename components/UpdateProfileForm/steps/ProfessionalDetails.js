import React from 'react';
import { bool } from 'prop-types';
import { Field } from 'formik';
import * as Yup from 'yup';
import { updateUser } from 'common/constants/api';
import { validationErrorMessages } from 'common/constants/messages';
import Input from 'components/Form/Input/Input';
import Select from 'components/Form/Select/Select';
import styles from './_steps.css';

ProfessionalDetails.propTypes = {
  isSubmitting: bool,
};

ProfessionalDetails.defaultProps = {
  isSubmitting: false,
};

ProfessionalDetails.title = 'Professional Details';

ProfessionalDetails.validationSchema = Yup.object().shape({
  employmentStatus: Yup.string()
    .nullable()
    .required(validationErrorMessages.required),
  companyName: Yup.string().nullable(),
  companyRole: Yup.string().nullable(),
});

ProfessionalDetails.initialValues = {
  employmentStatus: '',
  companyName: '',
  companyRole: '',
};

ProfessionalDetails.submitHandler = async values => {
  await updateUser(values);
};

function ProfessionalDetails({ isSubmitting }) {
  return (
    <>
      <div className={styles.row}>
        <Field
          className={styles.fullWidth}
          name="employmentStatus"
          label="Employment Status*"
          component={Select}
          options={[
            {
              value: 'fulltime',
              label: 'Employed Full-Time',
            },
            {
              value: 'parttime',
              label: 'Employed Part-Time',
            },
            {
              value: 'unemployed',
              label: 'Currently Unemployed',
            },
          ]}
          disabled={isSubmitting}
        />

        <Field
          type="text"
          name="companyName"
          label="Company Name"
          component={Input}
          disabled={isSubmitting}
        />

        <Field
          type="text"
          name="companyRole"
          label="Company Role"
          component={Input}
          disabled={isSubmitting}
        />
      </div>
    </>
  );
}

export default ProfessionalDetails;
