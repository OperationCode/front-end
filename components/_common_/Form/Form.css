.form {
  display: flex;
  flex-direction: column;
  width: 300px;
  max-width: 100%;
  margin: 0 auto 2rem;
}

.form * {
  box-sizing: border-box;
}

.formControl ~ .formControl {
  margin-top: 16px;
}

.formControl input {
  background-color: #e6e6e6;
  color: #000000;
  border: none;
  font-size: 1.3em;
  padding: 0 16px;
  height: 50px;
  width: 100%;
}

.formControl input::placeholder {
  color: #000000;
}
