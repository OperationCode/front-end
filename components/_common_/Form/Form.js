import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './Form.css';

Form.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.element]).isRequired,
  className: PropTypes.string,
};

Form.defaultProps = { className: '' };

function Form({ children, className }) {
  return (
    <form
      className={classNames({
        [`${styles.form}`]: true,
        [`${className}`]: className,
      })}
    >
      {children}
    </form>
  );
}

export default Form;
