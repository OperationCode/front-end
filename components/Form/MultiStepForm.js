/* eslint-disable unicorn/prevent-abbreviations */
import React, { useState } from 'react';
import { arrayOf, func, object } from 'prop-types';
import noop from 'lodash/noop';
import { Formik } from 'formik';
import { validStep } from 'common/constants/custom-props';
import Button from 'components/Button/Button';
import Form from 'components/Form/Form';
import Alert from 'components/Alert/Alert';
import ProgressIndicator from 'components/ProgressIndicator/ProgressIndicator';
import styles from './MultiStepForm.css';

MultiStepForm.propTypes = {
  // initialValues must be object where entire form's shape is described
  initialValues: object.isRequired,

  getErrorMessage: func.isRequired,
  onEachStepSubmit: func,
  onFinalSubmit: func.isRequired, // to be considered onSuccess
  steps: arrayOf(validStep).isRequired,
};

MultiStepForm.defaultProps = {
  onEachStepSubmit: noop,
};

export function MultiStepForm({
  steps,
  initialValues,
  onEachStepSubmit,
  onFinalSubmit,
  getErrorMessage,
}) {
  const [state, setState] = useState({
    errorMessage: '',
    stepNumber: 0,
  });

  const { errorMessage, stepNumber } = state;

  const isLastStep = () => {
    return stepNumber === steps.length - 1;
  };

  // We assume this method cannot be called on the last step
  const showNextStep = ({ setFieldTouched }) => {
    // TODO: Only untouch if value is '' or []
    const nextStepFieldNames = Object.keys(steps[stepNumber + 1].initialValues);
    nextStepFieldNames.forEach(fieldName => setFieldTouched(fieldName, false));
    setState(prevState => ({ ...prevState, stepNumber: prevState.stepNumber + 1 }));
  };

  // We assume this method cannot be called on the first step
  const showPreviousStep = ({ setFieldTouched }) => {
    // TODO: Only untouch if value is '' or []
    const previousStepFieldNames = Object.keys(steps[stepNumber - 1].initialValues);
    previousStepFieldNames.forEach(fieldName => setFieldTouched(fieldName, false));

    setState(prevState => ({ ...prevState, stepNumber: prevState.stepNumber - 1 }));
  };

  const handleError = error => {
    setState(prevState => ({ ...prevState, errorMessage: getErrorMessage(error) }));
  };

  const handleSubmit = async (values, formikBag) => {
    if (errorMessage) {
      // reset error message each submit
      setState(prevState => ({ ...prevState, errorMessage: '' }));
    }

    try {
      await onEachStepSubmit(values);

      const currentStepSubmitHandler = steps[stepNumber].submitHandler;
      await currentStepSubmitHandler(values);

      if (isLastStep()) {
        await onFinalSubmit(values);
        formikBag.setSubmitting(false);
        formikBag.resetForm();
      } else {
        formikBag.setSubmitting(false);
        showNextStep(formikBag);
      }
    } catch (error) {
      formikBag.setSubmitting(false);
      handleError(error);
    }
  };

  const CurrentStep = steps[stepNumber];
  const isFirstStep = stepNumber === 0;

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={CurrentStep.validationSchema}
      onSubmit={handleSubmit}
      render={formikBag => (
        <Form className={styles.MultiStepForm} onSubmit={formikBag.handleSubmit}>
          <h3 className={styles.centerAligned}>{CurrentStep.title}</h3>

          <ProgressIndicator stepNumber={stepNumber} totalSteps={steps.length} />

          <CurrentStep {...formikBag} />

          <div className={styles.errorMessage}>
            {errorMessage && <Alert type="error">{errorMessage}</Alert>}
          </div>

          <div className={styles.buttonGrouping}>
            {!isFirstStep && (
              <Button
                theme="secondary"
                disabled={formikBag.isSubmitting}
                onClick={() => showPreviousStep(formikBag)}
                data-testid="Previous Step Button"
              >
                ← Previous
              </Button>
            )}

            {isLastStep() ? (
              <Button
                type="submit"
                theme="secondary"
                disabled={formikBag.isSubmitting}
                data-testid="Submit Multi-Step Form"
              >
                Submit ✓
              </Button>
            ) : (
              <Button
                type="submit"
                theme="secondary"
                disabled={formikBag.isSubmitting}
                fullWidth={isFirstStep}
                data-testid="Submit Step Button"
              >
                Next →
              </Button>
            )}
          </div>
        </Form>
      )}
    />
  );
}

export default MultiStepForm;
