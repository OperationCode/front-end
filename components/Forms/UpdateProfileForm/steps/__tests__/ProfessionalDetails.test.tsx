import { Formik } from 'formik';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import Form from 'components/Form/Form';

import noop from 'lodash/noop';
import { ProfessionalDetails } from '../ProfessionalDetails';

describe('UpdateProfileForm/Steps/ProfessionalDetails', () => {
  it('should render in context of Formik', () => {
    createSnapshotTest(
      <Formik
        initialValues={ProfessionalDetails.initialValues}
        validationSchema={ProfessionalDetails.validationSchema}
        onSubmit={noop}
      >
        <Form>
          <ProfessionalDetails isSubmitting={false} />
        </Form>
      </Formik>,
    );
  });
});
