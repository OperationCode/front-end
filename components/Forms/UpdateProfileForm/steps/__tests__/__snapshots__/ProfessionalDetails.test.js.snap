// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpdateProfileForm/Steps/ProfessionalDetails > should render in context of Formik 1`] = `
<form
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <input
      className="fullWidth"
      label="Employment Status*"
      name="employmentStatus"
      onBlur={[Function]}
      onChange={[Function]}
      options={
        [
          {
            "label": "Employed Full-Time",
            "value": "Full-Time",
          },
          {
            "label": "Employed Part-Time",
            "value": "Part-Time",
          },
          {
            "label": "Currently Unemployed",
            "value": "Unemployed",
          },
        ]
      }
      value=""
    />
    <div
      className="field"
      data-testid="INPUT"
    >
      <label
        className="text-themeSecondary text-sm"
        data-testid="LABEL"
        htmlFor="companyName"
      >
        Company Name
      </label>
      <div
        className="inputFeedbackGrouping"
        data-testid="INPUT_FEEDBACK_GROUPING"
      >
        <input
          className="Input"
          id="companyName"
          name="companyName"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
      data-testid="INPUT"
    >
      <label
        className="text-themeSecondary text-sm"
        data-testid="LABEL"
        htmlFor="companyRole"
      >
        Company Role
      </label>
      <div
        className="inputFeedbackGrouping"
        data-testid="INPUT_FEEDBACK_GROUPING"
      >
        <input
          className="Input"
          id="companyRole"
          name="companyRole"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</form>
`;
